lib/bolt/transpiler.rb

```diff
@@ -0,0 +1,125 @@
+
+        # Param name
+        plan_string << "$#{param.name}"
+
+        # Param default
+        if param.value
+          default = type =~ /String/ ? "'#{param.value}'" : param.value
+          plan_string << " = #{default}"
+        end
+
+        # If it's the last parameter add a newline and no comma
+        last = i + 1 == plan_object.parameters.length ? "\n" : ","
+        # This encodes strangely if we << directly to plan_string
+        plan_string << last
+      end
+      plan_string << ") {\n"
+
+      plan_object.steps&.each do |step|
+        step.transform_values! { |v| plan_object.stringify(v) }
+        plan_string << "$#{step['name']} = " if step['name']
+        step_type, *extra_keys = Bolt::PAL::YamlPlan::Evaluator::STEP_KEYS.select { |key| step.key?(key) }
+
+        if !step_type || extra_keys.any?
+          raise Bolt::ConvertError.new("Can't convert unsupported step type #{step.name}", @plan_path)
+        end
+
+        target = step['target']
+        description = step.fetch('description', nil)
+        parameters = step.fetch('parameters', {})
+        if step_type == 'script' && step.key?('arguments')
+          parameters['arguments'] = step['arguments']
+        end
+
+        case step_type
+        when 'command', 'task', 'script', 'plan'
+          plan_string << "run_#{step_type}('#{step[step_type]}', #{target}"
+          plan_string << ", '#{description}'" if description
+          plan_string << ", #{parameters}" unless parameters.empty?
+        when 'source'
+          plan_string << "upload_file('#{step['source']}', '#{step['destination']}'"
+          plan_string << ", '#{description}'" if description
+        when 'eval'
+          # TODO
+        else
+          # We should never get here
+          raise Bolt::ConvertError.new("Can't convert unsupported step type #{step.name}", @plan_path)
+        end
+        plan_string << ")"
+      end
+
+      plan_string << "\nreturn #{plan_object.stringify(plan_object.return)}" if plan_object.return
+      plan_string << "\n}"
+      plan_string
+    end
+
+    def self.parse_plan
+      file_contents = File.read(@plan_path)
+      begin
+        result = Bolt::PAL::YamlPlan::Loader.parse_plan(file_contents, @plan_path)
+      rescue Error => e
+        raise Bolt::ConvertError.new("Failed to convert yaml plan: #{e.message}", @plan_path)
+      end
+
+      unless result.is_a?(Hash)
+        type = result.class.name
+        raise ArgumentError, "The data loaded from #{source_ref} does not contain an object - its type is #{type}"
+      end
+
+      # TODO: probably a better way to do this?
+      basename = File.basename(@filename, '.yaml')
+      fake_typed_name = if basename == 'init'
+                          @module
+                        else
+                          [@module, File.basename(@filename, '.yaml')].join('::')
+                        end
+      Bolt::PAL::YamlPlan.new(fake_typed_name, result).freeze
```

Comment:
After https://github.com/puppetlabs/bolt/pull/954 I think some of the validation in the `Transpiler` code will be redundant. (for example random step keys, etc).

Reply:
YAY awesome. I kind of assumed *most* of the checking would be done there (like that targets is in every step).